File: android/app/src/main/kotlin/fyi/yours/app/net/SseParser.kt
package fyi.yours.app.net

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

// One server-sent event off the /stream wire. The server relays upstream
// event names and adds its own (universe_time, error, end). See PROTOCOL.md
// for framing details.
data class SseEvent(val name: String, val data: String?) {
    private val json: JsonObject? by lazy {
        data?.let { runCatching { Json.parseToJsonElement(it).jsonObject }.getOrNull() }
    }

    val textDelta: String?
        get() {
            if (name != "content_block_delta") return null
            val delta = json?.get("delta")?.jsonObject ?: return null
            if (delta["type"]?.jsonPrimitive?.content != "text_delta") return null
            return delta["text"]?.jsonPrimitive?.content
        }

    val universeTime: String?
        get() {
            if (name != "universe_time") return null
            return json?.get("universe_time")?.jsonPrimitive?.content
        }

    val errorMessage: String?
        get() {
            if (name != "error") return null
            return json?.get("error")?.jsonObject?.get("message")?.jsonPrimitive?.content
        }
}

// Line-by-line SSE parser. Dispatch-on-blank-line per the SSE spec, with one
// accommodation for this server: the final "end" event arrives without a
// trailing blank line, so finish() flushes whatever is pending at EOF.
class SseLineParser {
    private var pendingName: String? = null
    private var pendingData: String? = null

    fun consume(line: String): SseEvent? {
        if (line.isEmpty()) return flush()
        if (line.startsWith("event:")) {
            val flushed = if (pendingData != null) flush() else null
            pendingName = line.removePrefix("event:").trim()
            return flushed
        }
        if (line.startsWith("data:")) {
            val value = line.removePrefix("data:").trim()
            pendingData = pendingData?.let { "$it\n$value" } ?: value
            return null
        }
        return null
    }

    fun finish(): SseEvent? = flush()

    private fun flush(): SseEvent? {
        if (pendingName == null && pendingData == null) return null
        val event = SseEvent(pendingName ?: "message", pendingData)
        pendingName = null
        pendingData = null
        return event
    }
}

File: ios/Yours/Networking/SSEParser.swift
import Foundation

// One server-sent event off the /stream wire. The server relays upstream
// event names and adds its own (universe_time, error, end). See PROTOCOL.md
// for framing details.
struct SSEEvent: Equatable {
    var name: String
    var data: String?

    // Decodes the data payload's interesting fields without committing to a
    // full schema — unknown events pass through harmlessly.
    var textDelta: String? {
        guard name == "content_block_delta", let object = jsonObject,
              let delta = object["delta"] as? [String: Any],
              delta["type"] as? String == "text_delta"
        else { return nil }
        return delta["text"] as? String
    }

    var universeTime: String? {
        guard name == "universe_time", let object = jsonObject else { return nil }
        return object["universe_time"] as? String
    }

    var errorMessage: String? {
        guard name == "error", let object = jsonObject,
              let error = object["error"] as? [String: Any]
        else { return nil }
        return error["message"] as? String
    }

    private var jsonObject: [String: Any]? {
        guard let data = data?.data(using: .utf8) else { return nil }
        return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
    }
}

// Line-by-line SSE parser. Dispatch-on-blank-line per the SSE spec, with one
// accommodation for this server: the final "end" event arrives without a
// trailing blank line, so finish() flushes whatever is pending at EOF.
struct SSELineParser {
    private var pendingName: String?
    private var pendingData: String?

    mutating func consume(line: String) -> SSEEvent? {
        if line.isEmpty {
            return flush()
        }
        if line.hasPrefix("event:") {
            let flushed = (pendingData != nil) ? flush() : nil
            pendingName = String(line.dropFirst("event:".count)).trimmingCharacters(in: .whitespaces)
            return flushed
        }
        if line.hasPrefix("data:") {
            let value = String(line.dropFirst("data:".count)).trimmingCharacters(in: .whitespaces)
            pendingData = pendingData.map { $0 + "\n" + value } ?? value
            return nil
        }
        return nil
    }

    mutating func finish() -> SSEEvent? {
        flush()
    }

    private mutating func flush() -> SSEEvent? {
        guard pendingName != nil || pendingData != nil else { return nil }
        let event = SSEEvent(name: pendingName ?? "message", data: pendingData)
        pendingName = nil
        pendingData = nil
        return event
    }
}
